### PR TITLE
fix: support existential type

### PR DESCRIPTION
**Pull request by csmoe**
Closes https://github.com/rust-analyzer/rust-analyzer/issues/1387
r? @matklad

### CODE REVIEW

File: crates/ra_parser/src/grammar/items.rs [hunk 1]
```diff
@@ -107,6 +107,10 @@ pub(super) fn maybe_item(p: &mut Parser, m: Marker, flavor: ItemFlavor) -> Resul
         p.bump_remap(T![default]);
         has_mods = true;
     }
+    if p.at(IDENT) && p.at_contextual_kw("existential") && p.nth(1) == T![existential] {
```
Comment (matklad) [inline crates/ra_parser/src/grammar/items.rs]:
```suggestion
    if p.at(IDENT) && p.at_contextual_kw("existential") && p.nth(1) == T![type] {
```

I think we should look ahead for `type` here

File: crates/ra_syntax/src/grammar.ron [hunk 2]
```diff
@@ -153,6 +154,7 @@ Grammar(
         "FOR_TYPE",
         "IMPL_TRAIT_TYPE",
         "DYN_TRAIT_TYPE",
+        "EXISTENTIAL_TYPE",
```
Comment (matklad) [inline crates/ra_syntax/src/grammar.ron]: Ah, one more thing!

I don't think we need to create a separate `EXISTENTIAL_TYPE` node type. Rather, we should re-use existing `TYPE_ALIAS`, and treat existential as a modifer. So, `type_alias_def` should still call `m.complete(p, TYPE_ALIAS_DEF);` inside

Comment (csmoe) [reply crates/ra_syntax/src/grammar.ron]: when refers to something `alias`, `=` is a must. I am a bit concern to treat `EXISTENTIAL_TYPE` as a `special` case of `alias`.

Comment (matklad) [reply crates/ra_syntax/src/grammar.ron]: I think that's OK: we already don't require `=` in types in traits:

```
trait Future {
    type Output;
}
```

Comment (csmoe) [reply crates/ra_syntax/src/grammar.ron]: okay, I'm persuaded :)

**Comment by matklad** [comment]
bors r+

Thanks!

**Pull request merged**